### PR TITLE
Feat: Add  dedicated Transaction Page.

### PR DESCRIPTION
**Pull request by abbasabid**
# Description of change

- Added a dedicated transaction page (for search by transactionId).
- Fix issues of `searchApi` in stardustTangleHelper.
- Add redirection to `Transaction` page by clicking on `transactionId` on `Block` page .

Aims to fix https://github.com/iotaledger/explorer/issues/329

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: client/src/app/App.tsx [hunk 1]
```diff
@@ -292,6 +295,13 @@ class App extends Component<RouteComponentProps<AppRouteProps> & { config: IConf
                                                     withNetworkProvider(<StardustBlock {...props} />)
                                                 )}
                                             />
+                                            <Route
```
Comment (msarcev) [inline client/src/app/App.tsx:298]: Please here add guard to also render if `isStardust` only, else on non-stardust networks we might have collision or routes.

File: api/src/utils/stardust/stardustTangleHelper.ts [hunk 2]
```diff
@@ -312,29 +338,31 @@ export class StardustTangleHelper {
...
+                if (Object.keys(transactionBlock).length > 0) {
+                    const writeStream = new WriteStream();
+                    serializeBlock(writeStream, transactionBlock);
+                    const includedBlockId = Converter.bytesToHex(Blake2b.sum256(writeStream.finalBytes()));
 
-                if (Object.keys(block).length > 0) {
                     return {
-                        block
+                        transactionBlock,
+                        includedBlockId
```
Comment (msarcev) [inline api/src/utils/stardust/stardustTangleHelper.ts:352]: Seams like we don't use this for anything. Let's remove it and related code above.

File: client/src/app/routes/stardust/TransactionPage.scss [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            }
+
+            .section {
+                padding-top: 44px;
+
+                .section--header {
+                    margin-top: 44px;
+                }
+            }
+
```
Comment (msarcev) [inline client/src/app/routes/stardust/TransactionPage.scss:51]: Remove extra vertical spacing.

File: client/src/app/routes/stardust/TransactionPage.tsx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            this.setState({
+                inputs,
+                outputs,
+                transferTotal,
+                block,
+                networkId: block.payload.essence.networkId,
+                inputsCommitment: block.payload.essence.inputsCommitment
+            });
+        } else {
+            this.props.history.replace(`/${this.props.match.params.network
```
Comment (msarcev) [inline client/src/app/routes/stardust/TransactionPage.tsx:77]: Please, no line break in the middle of these URL paths.

File: client/src/app/routes/stardust/TransactionPage.tsx [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+     */
+    constructor(props: RouteComponentProps<TransactionPageProps>) {
+        super(props);
+
+        this._tangleCacheService = ServiceFactory.get<StardustTangleCacheService>(
+            `tangle-cache-${STARDUST}`
+        );
+
+        this.state = {
+            transactionId: this.props.match.params.transactionId
```
Comment (msarcev) [inline client/src/app/routes/stardust/TransactionPage.tsx:45]: We don't need to put this in the state if we have it in the props, and it doesn't ever change.

File: client/src/models/api/stardust/ITransactionBlockRequest.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export interface ITransactionBlockRequest {
```
Comment (msarcev) [inline client/src/models/api/stardust/ITransactionBlockRequest.ts:1]: Can we rename it too `ITransactionDetailsRequest` ?

File: client/src/models/api/stardust/ITransactionBlockResponse.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { IBlock } from "@iota/iota.js-stardust";
```
Comment (msarcev) [inline client/src/models/api/stardust/ITransactionBlockResponse.ts:1]: Can we rename it too `ITransactionDetailsResponse` ?

File: client/src/models/api/stardust/ITransactionBlockResponse.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+import { IResponse } from "../IResponse";
+
+export interface ITransactionBlockResponse extends IResponse {
+    /**
+     * The transaction included block.
+     */
+    transactionBlock?: IBlock;
```
Comment (msarcev) [inline client/src/models/api/stardust/ITransactionBlockResponse.ts:8]: Rename field to simply `block`. Transaction block is confusing.

File: api/src/models/api/stardust/ITransactionBlockRequest.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+export interface ITransactionBlockRequest {
```
Comment (msarcev) [inline api/src/models/api/stardust/ITransactionBlockRequest.ts:1]: Rename these like described on client side.

**Comment by msarcev** [comment]
Also let's add a field "Included in Block" with `blockId` value, under General > "Transaction Id".

File: api/src/utils/stardust/stardustTangleHelper.ts (hunk 2 above)
Comment (abbasabid) [reply api/src/utils/stardust/stardustTangleHelper.ts]: Resolved.

File: client/src/app/routes/stardust/TransactionPage.tsx (hunk 5 above)
Comment (abbasabid) [reply client/src/app/routes/stardust/TransactionPage.tsx]: Resolved.

File: client/src/models/api/stardust/ITransactionBlockResponse.ts (hunk 8 above)
Comment (abbasabid) [reply client/src/models/api/stardust/ITransactionBlockResponse.ts]: Resolved.

**Comment by abbasabid** [comment]
> Also let's add a field "Included in Block" with `blockId` value, under General > "Transaction Id".

Added a function in `transactionHelper.ts` to compute blockId. Please have a look and share your feedback.

**Comment by msarcev** [comment]
> > Also let's add a field "Included in Block" with `blockId` value, under General > "Transaction Id".
> 
> Added a function in `transactionHelper.ts` to compute blockId. Please have a look and share your feedback.

Please use `TransactionHelper.calculateBlockId` from `iota.js-stardust`.

File: client/src/app/routes/stardust/TransactionPage.tsx [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+     */
+    constructor(props: RouteComponentProps<TransactionPageProps>) {
+        super(props);
+
+        this._tangleCacheService = ServiceFactory.get<StardustTangleCacheService>(
+            `tangle-cache-${STARDUST}`
+        );
+
+        this.state = {
+            networkId: "",
```
Comment (msarcev) [inline client/src/app/routes/stardust/TransactionPage.tsx:46]: No need to set state to empty string. They will be undefined on initialization.

Comment (msarcev) [reply client/src/app/routes/stardust/TransactionPage.tsx:46]: You can remove the whole `this.state =`.

**Review by msarcev — CHANGES_REQUESTED** [review]

**Comment by abbasabid** [comment]
> > > Also let's add a field "Included in Block" with `blockId` value, under General > "Transaction Id".
> > 
> > 
> > Added a function in `transactionHelper.ts` to compute blockId. Please have a look and share your feedback.
> 
> Please use `TransactionHelper.calculateBlockId` from `iota.js-stardust`.

Done

File: client/src/app/routes/stardust/Block.tsx [hunk 11]
```diff
@@ -458,14 +463,14 @@ class Block extends AsyncComponent<RouteComponentProps<BlockProps>, BlockState>
             });
 
             this.setState({
-                actualBlockId: result.includedBlockId ?? blockId,
```
[the remark this answers is not in the document]
Comment (abbasabid) [reply client/src/app/routes/stardust/Block.tsx:461]: Done.

File: client/src/app/routes/stardust/TransactionPageState.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
+import { IBlock, IUTXOInput } from "@iota/iota.js-stardust";
+import { IInput } from "../../../models/api/stardust/IInput";
+import { IOutput } from "../../../models/api/stardust/IOutput";
+
+export interface TransactionPageState {
+    /**
+     * The network id.
+     */
+    networkId?: string;
```
[the remark this answers is not in the document]
Comment (abbasabid) [reply client/src/app/routes/stardust/TransactionPageState.ts]: Done.

[the remark this answers is not in the document]
Comment (abbasabid) [reply client/src/app/routes/stardust/TransactionPageState.ts]: Done.

File: client/src/app/routes/stardust/TransactionPage.tsx (hunk 10 above)
Comment (abbasabid) [reply client/src/app/routes/stardust/TransactionPage.tsx]: Done.

**Comment by abbasabid** [comment]
**Transaction Page**
Screenshot 2022-07-01 at 4 54 08 PM
This is how the page looks. @luca-moser @lzpap any feedback?

**Comment by lzpap** [comment]
> **Transaction Page** Screenshot 2022-07-01 at 4 54 08 PM This is how the page looks. @luca-moser @lzpap any feedback?

The inclusion state definitely is missing. I can't tell from this page whether this is valid, not valid, confirmed , etc. Besides, it looks good.

So I suggest you load the metadata of the block of the tx and display that info here as well.

PS.: there might already be an issue for this, but in a future PR we should definitely imlement an "inline inputs" mode, where instead of just seing the ID of the consumed UTXO, it is actually fetched and displayed like the outputs on the other side. Without this, it is pretty hard to reason about where exactly the funds come from, since you only see the ID of the output you consume but not what is inside it. Anyway, it's just something for later.

**Comment by msarcev** [comment]
image

**Pull request merged**